### PR TITLE
[hostmetricsreceiver]: support reporting amount of available logical and physical CPUs

### PR DESCRIPTION
**Pull request by frzifus**
**Description:** <Describe what has changed.>

**Link to tracking Issue:** <Issue number if applicable>
- #22099 

**Testing:** <Describe what testing was performed and which tests were added.>
Manually:
```yaml
---
receivers:
  hostmetrics:
    collection_interval: 30s
    scrapers:
      cpu:
        metrics:
          system.cpu.time:
            enabled: false
          system.cpu.utilization:
            enabled: false
          system.cpu.count:
            enabled: true

exporters:
  logging:
    verbosity: detailed
  prometheus:
    endpoint: "127.0.0.1:9090"

service:
  pipelines:
    metrics:
      receivers: [hostmetrics]
      exporters: [logging, prometheus]
```

Got:
```
Resource SchemaURL: https://opentelemetry.io/schemas/1.9.0
ScopeMetrics #0
ScopeMetrics SchemaURL: 
InstrumentationScope otelcol/hostmetricsreceiver/cpu 0.79.0-dev
Metric #0
Descriptor:
     -> Name: system.cpu.count
     -> Description: Number of available CPUs.
     -> Unit: 1
     -> DataType: Gauge
NumberDataPoints #0
Data point attributes:
     -> machine-id: Str(1819479dc7174971967d4f2f1ba7b794)
StartTimestamp: 2023-06-08 13:40:33 +0000 UTC
Timestamp: 2023-06-08 14:11:56.118141589 +0000 UTC
Value: 8
	{"kind": "exporter", "data_type": "metrics", "name": "logging"}


---


# HELP system_cpu_count_ratio Nnumber of available CPUs.
# TYPE system_cpu_count_ratio gauge
system_cpu_count_ratio{machine_id="1819479dc7174971967d4f2f1ba7b794"} 8
```

**Documentation:** <Describe the documentation added.>

`documentation.md` was automatically updated by `mdatagen`.

### CODE REVIEW

File: receiver/hostmetricsreceiver/internal/scraper/cpuscraper/metadata.yaml [hunk 1]
```diff
@@ -30,3 +30,10 @@ metrics:
     gauge:
       value_type: double
     attributes: [cpu, state]
+
+  system.cpu.number:
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/cpuscraper/metadata.yaml:34]: This probably should be called `system.cpu.count`. See https://github.com/open-telemetry/semantic-conventions/blob/4456e4d1d4c2858ffd495a5cc6dc02f802dbaec4/specification/metrics/semantic_conventions/README.md#use-count-instead-of-pluralization

**Comment by dmitryax** [comment]
I'd be great if you can submit a PR to https://github.com/open-telemetry/semantic-conventions/blob/4456e4d1d4c2858ffd495a5cc6dc02f802dbaec4/specification/metrics/semantic_conventions/system-metrics.md#systemcpu---processor-metrics as well

File: receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_linux.go [hunk 2]
```diff
@@ -7,13 +7,19 @@
...
+
 	"github.com/shirou/gopsutil/v3/cpu"
 	"go.opentelemetry.io/collector/pdata/pcommon"
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata"
 	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/hostmetricsreceiver/internal/scraper/cpuscraper/ucal"
 )
 
+func (s *scraper) recordCPUNumberDataPoint(now pcommon.Timestamp) {
+	s.mb.RecordSystemCPUNumberDataPoint(now, float64(runtime.NumCPU()))
```
Comment (frzifus) [inline receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_linux.go:20]: Maybe that is also not the right way to parse it here. But I assume its not to hard to get this value from somewhere else.

https://github.com/open-telemetry/opentelemetry-specification/pull/2292
https://github.com/open-telemetry/opentelemetry-specification/pull/2292

Comment (dmitryax) [reply receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_linux.go:20]: Thanks for posting the links. Reading through https://github.com/open-telemetry/opentelemetry-specification/pull/2384, I'm more confident that the metric is a better option than a resource attribute. I'm good with accepting this PR. Still, I would like to hear back from @mx-psi and have another PR against the spec similar to https://github.com/open-telemetry/opentelemetry-specification/pull/2384 but from the collector perspective. The host metrics receiver can get an actual number of processors as opposed to the Java library.

Comment (mx-psi) [reply receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_linux.go]: Funny, reading open-telemetry/opentelemetry-specification#2384 made me more convinced that this should be a resource attribute :smile: Let's continue the discussion over at open-telemetry/semantic-conventions/pull/99 and mark this as 'on hold' until that one is resolved.

**Comment by mx-psi** [comment]
Blocked by https://github.com/open-telemetry/semantic-conventions/pull/99

File: .chloggen/hostmetricsreceiver_support_cpu_count.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+# you should instead start your pull request title with [chore] or use the "Skip Changelog" label.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: hostmetricsreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Report number of CPUs as metric.
```
Comment (mx-psi) [inline .chloggen/hostmetricsreceiver_support_cpu_count.yaml:12]: needs an update to say that we are reporting logical and physical number

File: receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_others.go [hunk 4]
```diff
@@ -27,3 +27,7 @@ func (s *scraper) recordCPUUtilization(now pcommon.Timestamp, cpuUtilization uca
 	s.mb.RecordSystemCPUUtilizationDataPoint(now, cpuUtilization.Idle, cpuUtilization.CPU, metadata.AttributeStateIdle)
 	s.mb.RecordSystemCPUUtilizationDataPoint(now, cpuUtilization.Irq, cpuUtilization.CPU, metadata.AttributeStateInterrupt)
 }
+
+func (s *scraper) recordCPUPhysicalCountDataPoint(now pcommon.Timestamp, count int) {}
```
Comment (mx-psi) [inline receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_others.go:31]: I think `cpu.Counts` is supported in most platforms, should we support this?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_others.go [hunk 5]
```diff
@@ -18,11 +18,11 @@ func (s *scraper) recordCPUTimeMetric(now pcommon.Timestamp, cpuTime *cpu.TimesS
 func (s *scraper) recordCPUUtilization(now pcommon.Timestamp, cpuUtilization ucal.CPUUtilization) {}
 
 func getProcessName(processHandle, string) (string, error) {
-	return nil, nil
+	return "", nil
 }
 
 func getProcessExecutable(processHandle) (string, error) {
-	return nil, nil
+	return "", nil
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_others.go:25]: Sorry for nit picking, but this is unrelated. Please submit another PR is needed

Comment (frzifus) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper_others.go]: Done
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24379

File: receiver/hostmetricsreceiver/internal/scraper/cpuscraper/documentation.md [hunk 6]
```diff
@@ -39,9 +39,25 @@ metrics:
     enabled: true
 ```
 
+### system.cpu.logical.count
+
+Number of available logical CPUs.
+
+| Unit | Metric Type | Value Type |
+| ---- | ----------- | ---------- |
+| cpu | Gauge | Int |
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/cpuscraper/documentation.md:48]: I believe the units should be like this according to the OTel spec
```suggestion
| {cpu} | Gauge | Int |
```

File: receiver/hostmetricsreceiver/internal/scraper/cpuscraper/metadata.yaml [hunk 7]
```diff
@@ -27,8 +27,22 @@ metrics:
...
     unit: 1
     gauge:
       value_type: double
     attributes: [cpu, state]
+
+  system.cpu.physical.count:
+    enabled: false
+    description: Number of available physical CPUs.
+    unit: cpu
+    gauge:
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/cpuscraper/metadata.yaml:40]: Also they should be non monotonic sums, not gauges

File: receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_linux.go [hunk 8]
```diff
@@ -35,3 +35,11 @@ func (s *scraper) recordCPUUtilization(now pcommon.Timestamp, cpuUtilization uca
...
 	s.mb.RecordSystemCPUUtilizationDataPoint(now, cpuUtilization.Iowait, cpuUtilization.CPU, metadata.AttributeStateWait)
 }
+
+func (s *scraper) recordCPUPhysicalCountDataPoint(now pcommon.Timestamp, count int) {
+	s.mb.RecordSystemCPUPhysicalCountDataPoint(now, int64(count))
+}
+
+func (s *scraper) recordCPULogicalCountDataPoint(now pcommon.Timestamp, count int) {
+	s.mb.RecordSystemCPULogicalCountDataPoint(now, int64(count))
+}
```
Comment (dmitryax) [inline receiver/hostmetricsreceiver/internal/scraper/cpuscraper/cpu_scraper_linux.go:45]: Why do we need to extract these functions if there are identical for linux and others? We can just use `s.mb.RecordSystemCPUPhysicalCountDataPoint(now, int64(count))` in cpu_scraper.go

**Comment by frzifus** [comment]
I assume we can pickup this again. Since the https://github.com/open-telemetry/semantic-conventions/pull/99 got merged.

File: .chloggen/hostmetricsreceiver_support_cpu_count.yaml [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Report  logical and physical number of CPUs as metric.
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: [22099]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Comment (mx-psi) [inline .chloggen/hostmetricsreceiver_support_cpu_count.yaml:20]:
```suggestion
subtext: | 
  Use the `system.cpu.logical.count::enabled` and `system.cpu.physical.count::enabled` flags to enable them
```

**Review by mx-psi — APPROVED** [review]
LGTM, needs a rebase

**Review by dmitryax — APPROVED** [review]
